README.md
# LetterAvatar

Gem for creating letter avatar from user's name (or any other strong / character).

Code extracted from [discourse](https://www.discourse.org/) source (thanks guys!) - I needed this functionality in three projects, so here's the gem! :-)

[![Gem Version](https://badge.fury.io/rb/letter_avatar.svg)](https://badge.fury.io/rb/letter_avatar)
[![Code Climate](https://codeclimate.com/github/ksz2k/letter_avatar/badges/gpa.svg)](https://codeclimate.com/github/ksz2k/letter_avatar)

## Examples

#### Google's Inbox Palette

<img src="https://cloud.githubusercontent.com/assets/5518/13031513/43eefa76-d30b-11e5-8f06-85f8eb2a4fb6.png" width="60" /> <img src="https://cloud.githubusercontent.com/assets/5518/13031514/43ef6d8a-d30b-11e5-9fbc-38ae526b56b3.png" width="60" /> <img src="https://cloud.githubusercontent.com/assets/5518/13031517/43f0da12-d30b-11e5-8fef-6c7daf235a54.png" width="60" /> <img src="https://cloud.githubusercontent.com/assets/5518/13031515/43f0568c-d30b-11e5-95c5-1653361d4443.png" width="60" /> <img src="https://cloud.githubusercontent.com/assets/5518/13031512/43eebcc8-d30b-11e5-9f95-0093bfadd182.png" width="60" /> <img src="https://cloud.githubusercontent.com/assets/5518/13031516/43f0d0bc-d30b-11e5-8822-f01a6a138ff8.png" width="60" /> <img src="https://cloud.githubusercontent.com/assets/5518/13031519/44382430-d30b-11e5-96e4-bcd7ce5eb155.png" width="60" /> <img src="https://cloud.githubusercontent.com/assets/5518/13031518/44378d04-d30b-11e5-9400-55ff46b94cbe.png" width="60" /> <img src="https://cloud.githubusercontent.com/assets/5518/13031521/443a03cc-d30b-11e5-8467-9592e9dbb2ae.png" width="60" /> <img src="https://cloud.githubusercontent.com/assets/5518/13031523/443badc6-d30b-11e5-9d72-45613018cab4.png" width="60" /> <img src="https://cloud.githubusercontent.com/assets/5518/13031520/44394e14-d30b-11e5-966c-2eada89295c9.png" width="60" /> <img src="https://cloud.githubusercontent.com/assets/5518/13031522/443a71fe-d30b-11e5-88f4-37d1fd220abb.png" width="60" /> <img src="https://cloud.githubusercontent.com/assets/5518/13031525/44752b1e-d30b-11e5-8290-ed8888055e64.png" width="60" /> <img src="https://cloud.githubusercontent.com/assets/5518/13031524/4471cef6-d30b-11e5-9f4c-004f993dd27b.png" width="60" /> <img src="https://cloud.githubusercontent.com/assets/5518/13031526/4475a990-d30b-11e5-8be3-c8f4482dee03.png" width="60" /> <img src="https://cloud.githubusercontent.com/assets/5518/13031527/44772482-d30b-11e5-92f0-b9190c312d70.png" width="60" /> <img src="https://cloud.githubusercontent.com/assets/5518/13031528/447804ce-d30b-11e5-8002-9424d5474ddb.png" width="60" />

## Installation

System requirements

```bash
$ sudo apt-get install -y imagemagick
```

Mac OS X

```bash
$ brew install imagemagick ghostscript
```

Add this line to your application's Gemfile:

```ruby
gem 'letter_avatar'
```

And then execute:

```bash
$ bundle
```

Or install it yourself as:

```bash
$ gem install letter_avatar
```

## Configuration

```ruby
LetterAvatar.setup do |config|
  config.fill_color        = 'rgba(255, 255, 255, 1)' # default is 'rgba(255, 255, 255, 0.65)'
  config.cache_base_path   = 'public/system/lets'     # default is 'public/system'
  config.colors_palette    = :iwanthue                # default is :google
  config.weight            = 500                      # default is 300
  config.annotate_position = '-0+10'                  # default is -0+5
  config.letters_count     = 2                        # default is 1
  config.pointsize         = 70                       # default is 140
end
```

#### Color palette

We have three color palettes implemented: `iwanthue`, `google` and `custom`.

Each of them have different colors, but the `iwanthue` also differently calculates the color for specified username.

The `google` selected will generate the same avatar for both, "Krzysiek" and "ksz2k" usernames given (both of them starts with letter "k"), but `iwanthue` will calculate it's md5 and then selects color, so there's huge chance that these usernames get different colors.

##### Custom palette definition

You can define your own `custom` palette:

```ruby
LetterAvatar.setup do |config|
  config.colors_palette = :custom
  config.custom_palette = [[120, 132, 205], [91, 149, 249], [72, 194, 249], [69, 208, 226]]
end
```

##### Custom font definition

You can define your own `custom` palette:

```ruby
LetterAvatar.setup do |config|
  config.font = File.join(File.expand_path('../../', File.dirname(__FILE__)), 'app/assets/fonts', 'font_name.ext')
end
```

## Usage

```ruby
LetterAvatar.generate 'ksz2k', 200
=> "public/system/letter_avatars/2/K/87_178_230/200.png"
```

### In your controllers / views

There's also helper for this. To use it, you need:

* in your helper (eg. `ApplicationHelper`) or controller:

  ```ruby
  include LetterAvatar::AvatarHelper
  ```

* and in your view:

  ```ruby
  letter_avatar_for('ksz2k', 200)
  => "public/system/letter_avatars/2/K/87_178_230/200.png"
  # or
  letter_avatar_url('ksz2k', 200)
  => "/system/letter_avatars/2/K/87_178_230/200.png"
  # or even
  letter_avatar_tag('ksz2k', 200, class: 'av')
  => "<img class=\"av\" alt=\"ksz2k\" src=\"/system/letter_avatars/2/K/87_178_230/200.png\" />"
  ```

### In your model

Say, you have a model `User` (which must have attribute or method `name`)

```ruby
require 'letter_avatar/has_avatar'

class User
  include LetterAvatar::HasAvatar
  ...

  def name
    'ksz2k'
  end
end
```

Then, in your views you can use:

```ruby
@user.avatar_path(200)
=> "public/system/letter_avatars/2/K/87_178_230/200.png"
# or
@user.avatar_url(200)
=> "/system/letter_avatars/2/K/87_178_230/200.png"
```

### Way to support non [a-z0-9] charsets

```rb
class User
  def username_for_avatar
    # Translate Chinese hanzi to pinyin
    # https://github.com/flyerhzm/chinese_pinyin
    Pinyin.t(self.username)
  end
end
```

Then you can get right avatar now:

```rb
letter_avatar_for(user.username_for_avatar, 200)
# or
letter_avatar_tag(user.username_for_avatar, 200, class: 'av')
```

## Contributing

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request


CHANGELOG.md
0.3.9
-----

- small minor fixes and cleanups.

0.3.8
-----

- added `LetterAvatar.font` config option.

0.3.7
-----

- added `LetterAvatar.custom_palette` config option.
- bumped `FULLSIZE` to 600.

0.3.6
-----

- added `LetterAvatar.letters_count` config option.
- added `LetterAvatar.pointsize` config option.

0.3.5
-----

- `LetterAvatar::HasAvatar` helper module cleanup
- `LetterAvatar.path_to_url` helper method

0.3.4
-----

- enhanced `LetterAvatar::HasAvatar` helper module

0.3.3
-----

- removed `-interpolate` imagemagick option - it's removed in IMv7

0.3.2
-----

- extend objects with `ActionView::Helpers::AssetTagHelper`

0.3.1
-----

- added `LetterAvatar.fill_color` config option.

0.3.0
-----

- Removed `activesupport` dependency.

0.2.0
-----

- Removed `posix-spawn` dependency, Open3 speed is ok.

0.1.11
------

- Windows support.

0.1.10
------

- `LetterAvatar::AvatarHelper` module tweaks
  - `letter_avatar_url` method, returning avatar proper url
  - `letter_avatar_tag` doeas not require `ActionView::Helpers::AssetTagHelper`

0.1.9
-----

- `LetterAvatar::Configuration` module, `mattr_accessor` (depending on `ActiveSupport`) removed

  fixes [#7](https://github.com/ksz2k/letter_avatar/issues/7)

0.1.5
-----

- Use [posix-spawn](https://github.com/rtomayko/posix-spawn) to instead of Kernel#exec to get better performance.

  > fork(2) calls slow down as the parent process uses more memory due to the need to copy page tables.


lib/letter_avatar/colors.rb
require 'digest' 

module LetterAvatar
  module Colors
    PALETTES = [:google, :iwanthue, :custom]

    GOOGLE_COLORS = [
      [226, 95, 81], # A
      [242, 96, 145], # B
      [187, 101, 202], # C
      [149, 114, 207], # D
      [120, 132, 205], # E
      [91, 149, 249], # F
      [72, 194, 249], # G
      [69, 208, 226], # H
      [72, 182, 172], # I
      [82, 188, 137], # J
      [155, 206, 95], # K
      [212, 227, 74], # L
      [254, 218, 16], # M
      [247, 192, 0], # N
      [255, 168, 0], # O
      [255, 138, 96], # P
      [194, 194, 194], # Q
      [143, 164, 175], # R
      [162, 136, 126], # S
      [163, 163, 163], # T
      [175, 181, 226], # U
      [179, 155, 221], # V
      [194, 194, 194], # W
      [124, 222, 235], # X
      [188, 170, 164], # Y
      [173, 214, 125] # Z
    ].freeze

    IWANTHUE_COLORS = [
      [198, 125, 40],
      [61, 155, 243],
      [74, 243, 75],
      [238, 89, 166],
      [52, 240, 224],
      [177, 156, 155],
      [240, 120, 145],
      [111, 154, 78],
      [237, 179, 245],
      [237, 101, 95],
      [89, 239, 155],
      [43, 254, 70],
      [163, 212, 245],
      [65, 152, 142],
      [165, 135, 246],
      [181, 166, 38],
      [187, 229, 206],
      [77, 164, 25],
      [179, 246, 101],
      [234, 93, 37],
      [225, 155, 115],
      [142, 140, 188],
      [223, 120, 140],
      [249, 174, 27],
      [244, 117, 225],
      [137, 141, 102],
      [75, 191, 146],
      [188, 239, 142],
      [164, 199, 145],
      [173, 120, 149],
      [59, 195, 89],
      [222, 198, 220],
      [68, 145, 187],
      [236, 204, 179],
      [159, 195, 72],
      [188, 121, 189],
      [166, 160, 85],
      [181, 233, 37],
      [236, 177, 85],
      [121, 147, 160],
      [234, 218, 110],
      [241, 157, 191],
      [62, 200, 234],
      [133, 243, 34],
      [88, 149, 110],
      [59, 228, 248],
      [183, 119, 118],
      [251, 195, 45],
      [113, 196, 122],
      [197, 115, 70],
      [80, 175, 187],
      [103, 231, 238],
      [240, 72, 133],
      [228, 149, 241],
      [180, 188, 159],
      [172, 132, 85],
      [180, 135, 251],
      [236, 194, 58],
      [217, 176, 109],
      [88, 244, 199],
      [186, 157, 239],
      [113, 230, 96],
      [206, 115, 165],
      [244, 178, 163],
      [230, 139, 26],
      [241, 125, 89],
      [83, 160, 66],
      [107, 190, 166],
      [197, 161, 210],
      [198, 203, 245],
      [238, 117, 19],
      [228, 119, 116],
      [131, 156, 41],
      [145, 178, 168],
      [139, 170, 220],
      [233, 95, 125],
      [87, 178, 230],
      [157, 200, 119],
      [237, 140, 76],
      [229, 185, 186],
      [144, 206, 212],
      [236, 209, 158],
      [185, 189, 79],
      [34, 208, 66],
      [84, 238, 129],
      [133, 140, 134],
      [67, 157, 94],
      [168, 179, 25],
      [140, 145, 240],
      [151, 241, 125],
      [67, 162, 107],
      [200, 156, 21],
      [169, 173, 189],
      [226, 116, 189],
      [133, 231, 191],
      [194, 161, 63],
      [241, 77, 99],
      [241, 217, 53],
      [123, 204, 105],
      [210, 201, 119],
      [229, 108, 155],
      [240, 91, 72],
      [187, 115, 210],
      [240, 163, 100],
      [178, 217, 57],
      [179, 135, 116],
      [204, 211, 24],
      [186, 135, 57],
      [223, 176, 135],
      [204, 148, 151],
      [116, 223, 50],
      [95, 195, 46],
      [123, 160, 236],
      [181, 172, 131],
      [142, 220, 202],
      [240, 140, 112],
      [172, 145, 164],
      [228, 124, 45],
      [135, 151, 243],
      [42, 205, 125],
      [192, 233, 116],
      [119, 170, 114],
      [158, 138, 26],
      [73, 190, 183],
      [185, 229, 243],
      [227, 107, 55],
      [196, 205, 202],
      [132, 143, 60],
      [233, 192, 237],
      [62, 150, 220],
      [205, 201, 141],
      [106, 140, 190],
      [161, 131, 205],
      [135, 134, 158],
      [198, 139, 81],
      [115, 171, 32],
      [101, 181, 67],
      [149, 137, 119],
      [37, 142, 183],
      [183, 130, 175],
      [168, 125, 133],
      [124, 142, 87],
      [236, 156, 171],
      [232, 194, 91],
      [219, 200, 69],
      [144, 219, 34],
      [219, 95, 187],
      [145, 154, 217],
      [165, 185, 100],
      [127, 238, 163],
      [224, 178, 198],
      [119, 153, 120],
      [124, 212, 92],
      [172, 161, 105],
      [231, 155, 135],
      [157, 132, 101],
      [122, 185, 146],
      [53, 166, 51],
      [70, 163, 90],
      [150, 190, 213],
      [210, 107, 60],
      [166, 152, 185],
      [159, 194, 159],
      [39, 141, 222],
      [202, 176, 161],
      [95, 140, 229],
      [168, 142, 87],
      [93, 170, 203],
      [159, 142, 54],
      [14, 168, 39],
      [94, 150, 149],
      [187, 206, 136],
      [157, 224, 166],
      [235, 158, 208],
      [109, 232, 216],
      [141, 201, 87],
      [208, 124, 118],
      [142, 125, 214],
      [19, 237, 174],
      [72, 219, 41],
      [234, 102, 111],
      [168, 142, 79],
      [188, 135, 35],
      [95, 155, 143],
      [148, 173, 116],
      [223, 112, 95],
      [228, 128, 236],
      [206, 114, 54],
      [195, 119, 88],
      [235, 140, 94],
      [235, 202, 125],
      [233, 155, 153],
      [214, 214, 238],
      [246, 200, 35],
      [151, 125, 171],
      [132, 145, 172],
      [131, 142, 118],
      [199, 126, 150],
      [61, 162, 123],
      [58, 176, 151],
      [215, 141, 69],
      [225, 154, 220],
      [220, 77, 167],
      [233, 161, 64],
      [130, 221, 137],
      [81, 191, 129],
      [169, 162, 140],
      [174, 177, 222],
      [236, 174, 47],
      [233, 188, 180],
      [69, 222, 172],
      [71, 232, 93],
      [118, 211, 238],
      [157, 224, 83],
      [218, 105, 73],
      [126, 169, 36]
    ].freeze

    def self.for(username)
      public_send("with_#{LetterAvatar.colors_palette}", username)
    end

    def self.with_iwanthue(username)
      iwanthue[
        ::Digest::MD5.hexdigest(username)[0...15].to_i(16) % iwanthue.length
      ]
    end

    def self.with_google(username)
      char = username[0].upcase

      if /[A-Z]/ =~ char
        # 65 is 'A' ord
        idx = char.ord - 65
        google[idx]
      elsif /[\d]/ =~ char
        google[char.to_i]
      else
        google[::Digest::MD5.hexdigest(username)[0...15].to_i(16) % google.length]
      end
    end

		def self.with_custom(username)
			custom_palette = LetterAvatar.custom_palette
			custom_palette[::Digest::MD5.hexdigest(username)[0...15].to_i(16) % custom_palette.length]
		end
		
		def self.valid_custom_palette?(palette)
			return false unless palette.is_a?(Array)
			return palette.all? do |color| 
				false unless color.is_a?(Array)
				color.all? { |i| i.is_a?(Integer) }
			end
		end

    # Colors form Google Inbox
    # https://inbox.google.com
    def self.google
      GOOGLE_COLORS
    end

    # palette of optimally disctinct colors
    # cf. http://tools.medialab.sciences-po.fr/iwanthue/index.php
    # parameters used:
    #   - H: 0 - 360
    #   - C: 0 - 2
    #   - L: 0.75 - 1.5
    def self.iwanthue
      IWANTHUE_COLORS
    end
  end
end
